File: code/TemplateStudioForUWP/Templates/Uwp/Features/BackgroundTask._VB/Param_ProjectName/Services/BackgroundTaskService.vb
Imports Windows.ApplicationModel.Background

Imports Param_RootNamespace.Activation
Imports Param_RootNamespace.BackgroundTasks
Imports Param_RootNamespace.Core.Helpers

Namespace Services
    Friend Class BackgroundTaskService
        Inherits ActivationHandler(Of BackgroundActivatedEventArgs)

        Public Shared ReadOnly Property BackgroundTasks As IEnumerable(Of BackgroundTask)
            Get
                Return BackgroundTaskInstances
            End Get
        End Property

        Private Shared ReadOnly BackgroundTaskInstances As IEnumerable(Of BackgroundTask) = CreateInstances()

        Public Async Function RegisterBackgroundTasksAsync() As Task
            BackgroundExecutionManager.RemoveAccess()
            Dim result = Await BackgroundExecutionManager.RequestAccessAsync()

            If result = BackgroundAccessStatus.DeniedBySystemPolicy OrElse result = BackgroundAccessStatus.DeniedByUser Then
                Return
            End If

            For Each task As BackgroundTask In BackgroundTasks
                task.Register()
            Next
        End Function

        Public Shared Function GetBackgroundTasksRegistration(Of T As BackgroundTask)() As BackgroundTaskRegistration
            If Not BackgroundTaskRegistration.AllTasks.Any(Function(t1) t1.Value.Name = GetType(T).Name) Then
                ' This condition should not be met. If it is it means the background task was not registered correctly.
                ' Please check CreateInstances to see if the background task was properly added to the BackgroundTasks property.
                Return Nothing
            End If

            Return DirectCast(BackgroundTaskRegistration.AllTasks.FirstOrDefault(Function(t2) t2.Value.Name = GetType(T).Name).Value, BackgroundTaskRegistration)
        End Function

        Public Sub Start(taskInstance As IBackgroundTaskInstance)
            Dim task As BackgroundTask = BackgroundTasks.FirstOrDefault(Function(b) b.Match(taskInstance?.Task?.Name))

            If task Is Nothing Then
                ' This condition should not be met. It is it it means the background task to start was not found in the background tasks managed by this service.
                ' Please check CreateInstances to see if the background task was properly added to the BackgroundTasks property.
                Return
            End If

            task.RunAsync(taskInstance).FireAndForget()
        End Sub

        Protected Overrides Async Function HandleInternalAsync(args As BackgroundActivatedEventArgs) As Task
            Start(args.TaskInstance)

            Await Task.CompletedTask
        End Function

        Private Shared Function CreateInstances() As IEnumerable(Of BackgroundTask)
            Dim backgroundTasks As List(Of BackgroundTask) = New List(Of BackgroundTask)()

            Return backgroundTasks
        End Function
    End Class
End Namespace

File: code/TemplateStudioForUWP/Templates/Ft/BgTask/Param_ProjectName/Services/BackgroundTaskService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

using Windows.ApplicationModel.Activation;
using Windows.ApplicationModel.Background;

using Param_RootNamespace.Activation;
using Param_RootNamespace.BackgroundTasks;
using Param_RootNamespace.Core.Helpers;

namespace Param_RootNamespace.Services
{
    internal class BackgroundTaskService : ActivationHandler<BackgroundActivatedEventArgs>
    {
        public static IEnumerable<BackgroundTask> BackgroundTasks => BackgroundTaskInstances.Value;

        private static readonly Lazy<IEnumerable<BackgroundTask>> BackgroundTaskInstances =
            new Lazy<IEnumerable<BackgroundTask>>(() => CreateInstances());

        public async Task RegisterBackgroundTasksAsync()
        {
            BackgroundExecutionManager.RemoveAccess();
            var result = await BackgroundExecutionManager.RequestAccessAsync();

            if (result == BackgroundAccessStatus.DeniedBySystemPolicy
                || result == BackgroundAccessStatus.DeniedByUser)
            {
                return;
            }

            foreach (var task in BackgroundTasks)
            {
                task.Register();
            }
        }

        public static BackgroundTaskRegistration GetBackgroundTasksRegistration<T>()
            where T : BackgroundTask
        {
            if (!BackgroundTaskRegistration.AllTasks.Any(t => t.Value.Name == typeof(T).Name))
            {
                // This condition should not be met. If it is it means the background task was not registered correctly.
                // Please check CreateInstances to see if the background task was properly added to the BackgroundTasks property.
                return null;
            }

            return (BackgroundTaskRegistration)BackgroundTaskRegistration.AllTasks.FirstOrDefault(t => t.Value.Name == typeof(T).Name).Value;
        }

        public void Start(IBackgroundTaskInstance taskInstance)
        {
            var task = BackgroundTasks.FirstOrDefault(b => b.Match(taskInstance?.Task?.Name));

            if (task == null)
            {
                // This condition should not be met. It is it it means the background task to start was not found in the background tasks managed by this service.
                // Please check CreateInstances to see if the background task was properly added to the BackgroundTasks property.
                return;
            }

            task.RunAsync(taskInstance).FireAndForget();
        }

        protected override async Task HandleInternalAsync(BackgroundActivatedEventArgs args)
        {
            Start(args.TaskInstance);

            await Task.CompletedTask;
        }

        private static IEnumerable<BackgroundTask> CreateInstances()
        {
            var backgroundTasks = new List<BackgroundTask>();

            return backgroundTasks;
        }
    }
}
